api/src/main/java/tennisclub/dto/CourtDto.java

```diff
@@ -48,4 +48,27 @@ public String getPreviewImageUrl() {
     public void setPreviewImageUrl(String previewImageUrl) {
         this.previewImageUrl = previewImageUrl;
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof CourtDto)) return false;
```

Comment:
Sorry for being an annoying reviewer 😅in the reference implementation, they stricly don't use `instanceof`, but compare the classes directly like `getClass() != o.getClass()` _(I think it makes sense - either it is the same class or not, you can't get anything else)_ - however, I don't think they ever mentioned this on the lectures or write about this somewhere on the Internet🤔.

Reply:
@Fuyunori, you are **by no means** annoying; thank you for catching my mistakes! Anyway:

- If I remember correctly, the lecture on persistence specifically instructed us to use `instanceof` rather than `getClass()` in our implementations of `equals()`. However, it was mentioned in the context of JPA entities rather than DTOs and might not apply here.
- In any case, if DTOs are never inherited from -- as we discussed earlier -- are `getClass()` and `instanceof` no pretty much equivalent?

Reply:
@paveltobias Yes, as you said, they just mentioned `equals` and `hashCode` in the context of entity classes 😕As I was not sure in this matter _(again 🤣😆)_, I tried to make a research on `instanceof` and `getClass()` - these links pretty much sums everything 😄 

https://stackoverflow.com/questions/4989818/instanceof-vs-getclass
https://stackoverflow.com/questions/596462/any-reason-to-prefer-getclass-over-instanceof-when-generating-equals

TLDR: you're right, we can actually use `instanceof`, but it is pointless because we don't actually inheritance. I would probably summarize it this way: if you don't use inheritance ➡️  use `getClass` by default, otherwise use `instanceof` 😃

Reply:
@Fuyunori, actually, the sources you have provided seem to suggest that the best solution would be to:
- use `instanceof`,
- mark the `equals()` method `final`.

That way, both:
1. the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) is maintained (the DTO can be safely replaced by its subclass without altering the method's behavior);
2. the fulfillment of the `equals()` contract is ensured (since `equals()` cannot be overridden, the symmetry of equals cannot be broken by subclasses).

This is, of course, still relevant only if the DTOs will ever be subclassed. By a developer, they should not be (as discussed earlier). But they **could** be _magically_ subclassed under the hood by some of the libraries/frameworks we use. Hibernate does that with the DAOs (as discussed in the lectures); e.g. Mockito, Dozer, etc. **could** do similar _magic_ with the DTOs.

Perhaps, I would stick with the `instanceof` for now (since it cannot break anything) and ask the tutor about his opinion on this [religious debate](https://stackoverflow.com/a/596885). :smile:

Reply:
@paveltobias that would probably be the best course of action _(to ask the seminar tutor)_ 😄 however I find the `final equals` quite exotic and unusual😆